### DISCUSSION TITLE
DTO와 Entity 매핑 방법 논의

### DISCUSSION THREAD

**Discussion by Seol-JY** [toduck-App/toduck-backend#12] [Polls]
현재 우리 프로젝트에서는 DTO와 Entity 간 매핑을 각각의 클래스 내부에서 처리하고 있습니다. 이 방식에는 장단점이 있으며, 다른 접근 방식들과 비교하여 가장 적합한 방법을 선택하고자 합니다. 제공된 코드는 단순 예시일 뿐, 실제 적용되는 코드가 아닙니다.

## 현재 상황

현재 우리의 코드는 다음과 같은 구조를 가지고 있습니다:

```java
// Entity
public class Routine {
    // fields...

    public static Routine of(User user, Social social, RoutineCreateRequest request) {
        // creation logic...
    }
}

// DTO
public class RoutineResponseDto {
    // fields...

    public static RoutineResponseDto of(Routine routine) {
        // mapping logic...
    }
}
```

### 현재 방식의 장단점

장점:
- 간단하고 직관적인 구조
- 관련 로직이 해당 클래스 내에 위치하여 찾기 쉬움
- 추가적인 클래스나 의존성 없이 구현 가능
- 작은 프로젝트나 단순한 도메인 모델에 적합

단점:
- Entity가 DTO에 의존하게 됨 (계층 간 의존성 위반)
- DTO가 Entity에 의존하게 됨 (계층 간 의존성 위반)
- 각 클래스의 책임이 불분명해짐 (단일 책임 원칙 위반)
- 테스트하기 어려울 수 있음
- 복잡한 매핑 로직의 경우 클래스가 비대해질 수 있음

### 1. 별도의 Mapper 클래스 사용

```java
@Component
public class RoutineMapper {
    public Routine toEntity(User user, Social social, RoutineCreateRequest request) {
        // mapping logic...
    }

    public RoutineResponseDto toDto(Routine routine) {
        // mapping logic...
    }
}
```

장점:
- 계층 간 의존성 문제 해결
- 매핑 로직을 한 곳에서 관리 가능
- 단일 책임 원칙 준수
- 테스트하기 쉬움

단점:
- 추가적인 클래스 생성으로 인한 복잡성 증가
- 보일러플레이트 코드 증가 가능성

### 2. 매핑 라이브러리 사용 (예: MapStruct)

```java
@Mapper(componentModel = "spring")
public interface RoutineMapper {
    Routine toEntity(User user, Social social, RoutineCreateRequest request);
    RoutineResponseDto toDto(Routine routine);
}
```

장점:
- 계층 간 의존성 문제 해결
- 보일러플레이트 코드 감소
- 타입 안전성 보장
- 성능이 우수함

단점:
- 추가 라이브러리 의존성 발생
- 복잡한 매핑의 경우 커스텀 로직 작성이 필요할 수 있음
- 러닝 커브가 있을 수 있음

### 3. Factory 패턴 사용

```java
@Component
public class RoutineFactory {
    private final UserRepository userRepository;
    private final SocialRepository socialRepository;

    public RoutineFactory(UserRepository userRepository, SocialRepository socialRepository) {
        this.userRepository = userRepository;
        this.socialRepository = socialRepository;
    }

    public Routine createRoutine(RoutineCreateRequest request, Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new EntityNotFoundException("User not found"));
        Social social = socialRepository.findByUserId(userId)
            .orElseThrow(() -> new EntityNotFoundException("Social not found"));
        
        return new Routine(user, social, request.getTitle(), request.getDescription());
    }
}

@Component
public class RoutineResponseDtoFactory {
    public RoutineResponseDto createDto(Routine routine) {
        // mapping logic...
    }
}
```

장점:
- 객체 생성 로직을 중앙화하여 관리 가능
- 계층 간 의존성 문제 해결
- 복잡한 생성 로직 캡슐화
- 테스트하기 쉬움
- 다양한 생성 방식을 제공할 수 있음 (예: createRoutineWithDefaultSocial, createUrgentRoutine 등)

단점:
- 추가적인 클래스들로 인한 복잡성 증가
- Factory 클래스들이 비대해질 수 있음
- 단순한 매핑에는 과도한 솔루션일 수 있음

## Mapper vs Factory

- Mapper는 주로 단순 데이터 변환에 초점을 맞추며, Factory는 복잡한 객체 생성 과정을 추상화합니다.
- Mapper는 일반적으로 양방향 변환을 지원하고 상태를 가지지 않지만, Factory는 단방향 객체 생성에 집중하고 때로는 상태(의존성)를 가질 수 있습니다.
- 단순 데이터 변환만 필요하다면 Mapper가, 복잡한 생성 로직이나 의존성 관리가 필요하다면 Factory가 더 적합할 수 있습니다.

## 고려사항

- 코드의 가독성과 유지보수성
- 성능
- 개발 생산성
- 테스트 용이성
- 프로젝트의 복잡성과 규모

 각 접근 방식에 대한 실제 경험이 있다면 공유해 주시면 감사하겠습니다.

**Comment by kang20** [comment]
질문
1. Mapper 방식의 보일러플레이트 코드가 발생할 가능성이 증가한다 하셨는데 예를 들면 어떤 코드들인가요?
2. 2번 방식의 Mapper는 1번 방식의 Mapper처럼 라이브러리가 만들어주는건가요?
3. Factory 방식은 오로지 객체 생성만의 역할을 맡는건가요 아니면 다른 영속성 계층의 역할을 맡는건가요

의견

저는 만약 Factory 방식이 오로지 객체 생성만의 역할을 맡는다면 Mapper 방식이 낫다고 생각합니다.

 이유는 현재 저희 프로젝트는 service 계층에서 이미 단일 책임 원칙을 위반한다고 생각합니다 (도메인 로직과 영속성 코드가 동시에 있음)
하지만 저희 프로젝트는 타협점을 봤고 Service 코드에 영속성 코드가 있는걸 허락했습니다. 

 그런데 위처럼 Factory 방식이 객체 생성 역할만을 한다면 영속성 관련 코드가 Factory와 Service 두 계층에 걸친다고 생각해서 유지보수 비용이 증가한다고 생각합니다.
그래서 저희 프로젝트에서 entity 와 dto 변환에 의존성을 줄이는게 목적이면 Mapper 가 적당하다고 전 생각합니다.

하지만 Factory 방식에 다른 영속성 코드(엔티티간의 의존관계 맺거나하는 JPA 코드들) 이 들어가게 된다면  저희 UseCase도 사용하는김에 아예 핵사고날 같은 클린 아키텍처 도입도 고려해볼 필요 있다고 생각합니다

**Reply by Seol-JY** [reply]
@kang20 

> 1. Mapper 방식의 보일러플레이트 코드가 발생할 가능성이 증가한다 하셨는데 예를 들면 어떤 코드들인가요?

```java
@Component
public class RoutineMapper {
    public Routine toEntity(RoutineDto dto) {
        Routine routine = new Routine();
        routine.setTitle(dto.getTitle());
        routine.setDescription(dto.getDescription());
        routine.setStartDate(dto.getStartDate());
        routine.setEndDate(dto.getEndDate());
        routine.setPriority(dto.getPriority());
        routine.setStatus(dto.getStatus());
        // ... 더 많은 필드들 ...
        return routine;
    }

    public RoutineDto toDto(Routine entity) {
        RoutineDto dto = new RoutineDto();
        dto.setTitle(entity.getTitle());
        dto.setDescription(entity.getDescription());
        dto.setStartDate(entity.getStartDate());
        dto.setEndDate(entity.getEndDate());
        dto.setPriority(entity.getPriority());
        dto.setStatus(entity.getStatus());
        // ... 더 많은 필드들 ...
        return dto;
    }
}
```
위와 같은  코드를 계속 만들어야 한다는 의미입니다. 2번 방식에 비해서 번거롭다는 정도로만 이해하면 될 것 같습니다!

<br/>

> 2. 2번 방식의 Mapper는 1번 방식의 Mapper처럼 라이브러리가 만들어주는건가요?

맞습니다. 저도 존재만 알고 있고 실제로 사용해 본 적은 없는데, 필드명이 일치하는 dto <-> Entity 필드들을 별다른 보일러플레이트 코드 없이 자동으로 매핑을 진행하는 느낌으로 보입니다. 다만, 많은 객체가 생성에 참여하거나 보다 복잡한 생성 로직이 필요한 경우 커스텀이 필요하며, 러닝 커브가 존재한다는 단점이 있을 것으로 보입니다.  보다 자세한 `MapStruct` 사용 예는 아래 링크를 참고하세요.
https://blog.naver.com/n_cloudplatform/222957490406

<br/>

> 3. Factory 방식은 오로지 객체 생성만의 역할을 맡는건가요 아니면 다른 영속성 계층의 역할을 맡는건가요.

Factory가 직접적으로 영속성 계층의 역할을 맡지는 않지만, 영속성 계층과 연관된 작업을 수행할 수 있습니다. 예를 들어:

```java
@Component
public class RoutineFactory {
    private final UserRepository userRepository;
    private final SocialRepository socialRepository;

    @Autowired
    public RoutineFactory(UserRepository userRepository, SocialRepository socialRepository) {
        this.userRepository = userRepository;
        this.socialRepository = socialRepository;
    }

    public Routine createRoutine(RoutineCreateRequest request, Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new EntityNotFoundException("User not found"));
        Social social = socialRepository.findByUserId(userId)
            .orElseThrow(() -> new EntityNotFoundException("Social not found"));
        
        Routine routine = new Routine();
        routine.setUser(user);
        routine.setSocial(social);
        routine.setTitle(request.getTitle());
        routine.setDescription(request.getDescription());
        // ... 기타 설정 ...

        return routine;
    }
}
```
이 예시에서 Factory는 객체 생성뿐만 아니라 필요한 관련 엔티티를 조회하는 역할도 수행합니다. 그러나 실제 저장(save) 작업은 서비스 계층이나 리포지토리에서 수행하게 됩니다.

개인적으로 이방식은 계층 간 경계가 모호해진다는 생각이 드네요.

**Reply by kang20** [reply]
답변 이해했습니다.

저는 Mapper 1,2 번 방식 중 아무거나 상관없습니다. 

다만 저도 Factory 방식은 위 예제처럼 적용한다면 오히려 프로젝트 복잡도가 증가할 거 같은 느낌이 드네요

Mapper 방식 중에 2번이 러닝 커브가 올 수 있어도 적용 해보는거에 저는 긍정적입니다

**Comment by Seol-JY** [comment]
@kang20 @Junad-Park 

추가적으로 `MapStruct`와 `Mapper` 클래스 사용에 대한 부정적인 견해에 대해서도 참고하면 좋을 것 같습니다.
https://kth990303.tistory.com/403 (댓글 내용도 확인해보세요.)

생산성 측면에서는 오히려 마이너스일 수도 있겠네요.

**Comment by Junad-Park** [comment]
조금 번거로울지라도 `Mapper` 클래스를 사용하여 매핑하는데 한 표 던지겠습니다!

@Seol-JY 님이 올려주신 `MapStruct`에 대한 부정적인 견해를 확인해보았는데, 생산성이 크지 않은데 테스트 코드에서 에러가 날 수 있다는 점을 감안하면 매력적이지 못한 것 같습니다.

**Comment by Seol-JY** [comment]
`MapStruct`와 `Mapper` 클래스를 사용하는 경우에, 변환을 `Usecase`단에서 할지, `Service`단에서 할지, 아니면 두 레이어 모두에서 일어날 수 있는지 등에 대해서도 논의해봐야 할 것 같아요.

**Comment by kang20** [comment]
MapStruct 라이브러리 오류점 있다는거 보니까 Mapper 1번으로 투표바꾸겠습니다. 그리고 Mapper 만약에 사용한다면 근데 정적 메서드 형태로도 구현할 수 있나요?

**Reply by Seol-JY** [reply]
@kang20 할 수는 있는데, 아마 `@Component` 로 Bean으로 띄워서 사용 할 것 같아요. 우선 개발할 때 반영해보겠습니다.
